src/Dictionaries/RegExpTreeDictionary.cpp

```diff
@@ -0,0 +1,545 @@
+    {
+        Field field;
+        std::vector<StringPiece> pieces;
+
+        constexpr bool containsBackRefs() const { return !pieces.empty(); }
+    };
+
+    std::unordered_map<String, AttributeValue> attributes;
+};
+
+std::vector<StringPiece> createStringPieces(const String & value, int num_captures, const String & regex, Poco::Logger * logger)
+{
+    std::vector<StringPiece> result;
+    String literal;
+    for (size_t i = 0; i < value.size(); ++i)
+    {
+        if ((value[i] == '\\' || value[i] == '$') && i + 1 < value.size())
+        {
+            if (isNumericASCII(value[i+1]))
+            {
+                if (!literal.empty())
+                {
+                    result.push_back(StringPiece(literal));
+                    literal = "";
+                }
+                int ref_num = value[i+1]-'0';
+                if (ref_num >= num_captures)
+                    LOG_DEBUG(logger,
+                        "Reference Id {} in set string is invalid, the regexp {} only has {} capturing groups",
+                        ref_num, regex, num_captures-1);
+                result.push_back(StringPiece(ref_num));
+                ++i;
+                continue;
+            }
+        }
+        literal += value[i];
+    }
+    if (result.empty())
+        return result;
+    if (!literal.empty())
+        result.push_back(StringPiece(literal));
+    return result;
+}
+
+void RegExpTreeDictionary::calculateBytesAllocated()
+{
+    for (const String & regex : regexps)
+        bytes_allocated += regex.size();
+    bytes_allocated += sizeof(UInt64) * regexp_ids.size();
+    bytes_allocated += (sizeof(RegexTreeNode) + sizeof(UInt64)) * regex_nodes.size();
+    bytes_allocated += 2 * sizeof(UInt64) * topology_order.size();
+}
+
+void RegExpTreeDictionary::initRegexNodes(Block & block)
+{
+    auto id_column = block.getByName(kId).column;
+    auto pid_column = block.getByName(kParentId).column;
+    auto regex_column = block.getByName(kRegExp).column;
+    auto keys_column = block.getByName(kKeys).column;
+    auto values_column = block.getByName(kValues).column;
+
+    size_t size = block.rows();
+    for (size_t i = 0; i < size; i++)
+    {
+        UInt64 id = id_column->getUInt(i);
+        UInt64 parent_id = pid_column->getUInt(i);
+        String regex = (*regex_column)[i].safeGet<String>();
+
+        if (regex_nodes.contains(id))
+            throw Exception(ErrorCodes::INCORRECT_DICTIONARY_DEFINITION, "There are duplicate id {}", id);
+
+        if (id == 0)
+            throw Exception(ErrorCodes::INCORRECT_DICTIONARY_DEFINITION, "There are invalid id {}", id);
+
+        regexps.push_back(regex);
+        regexp_ids.push_back(id);
+
+        re2_st::RE2::Options regexp_options;
+        regexp_options.set_log_errors(false);
+        RegexTreeNodePtr node = std::make_unique<RegexTreeNode>(id, parent_id, regex, regexp_options);
+
+        int num_captures = std::min(node->searcher.NumberOfCapturingGroups() + 1, 10);
```

Comment:
Why 10 is the minimum?

Reply:
just copy from https://github.com/ClickHouse/ClickHouse/blob/150a699dda1033a35f267689ece8f81f1cc672e6/src/Functions/ReplaceRegexpImpl.h#L49
and in uap-core the back-ref is also $1-$9, so I think this restraint is reasonable.